### ISSUE TITLE
[18.01] sickle tool fails on list of pairs (due to structured_like)

### ISSUE THREAD

**Issue by nsoranzo** [galaxyproject/galaxy#6157]
On 17.01-17.09 the [sickle tool](https://github.com/galaxyproject/tools-iuc/blob/master/tools/sickle/sickle.xml) from iuc can be run on a list of paired FASTQs and produces another list of pairs, with the same structure and formats, using `structured_like="input_paired" inherit_format="true"`.
This was broken on 18.01 (in commit 78babab6455f4cd1c51aed0b1ddf6b4b29ad6d03, according to git bisect) and now fails with the following errors:

![screenshot from 2018-05-16 23-42-16](https://user-images.githubusercontent.com/4924623/40147835-2584d160-5963-11e8-9e53-cdad1ab25241.png)

```
127.0.0.1 - - [16/May/2018:23:42:03 +0100] "POST /api/tools/toolshed.g2.bx.psu.edu/repos/iuc/sickle/sickle/1.33.1/build HTTP/1.1" 200 - "http://127.0.0.1:8080/?tool_id=toolshed.g2.bx.psu.edu%2Frepos%2Fiuc%2Fsickle%2Fsickle%2F1.33.1&version=1.33.1&__identifer=eqdjaao7u85" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0"
galaxy.tools DEBUG 2018-05-16 23:42:05,300 Validated and populated state for tool request (43.218 ms)
galaxy.web.framework.decorators ERROR 2018-05-16 23:42:05,332 Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/opt/galaxy/lib/galaxy/web/framework/decorators.py", line 281, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/opt/galaxy/lib/galaxy/webapps/galaxy/api/tools.py", line 354, in create
    vars = tool.handle_input(trans, incoming, history=target_history, use_cached_job=use_cached_job)
  File "/opt/galaxy/lib/galaxy/tools/__init__.py", line 1337, in handle_input
    execution_tracker = execute_job(trans, self, mapping_params, history=request_context.history, rerun_remap_job_id=rerun_remap_job_id, collection_info=collection_info, completed_jobs=completed_jobs)
  File "/opt/galaxy/lib/galaxy/tools/execute.py", line 81, in execute
    execution_tracker.ensure_implicit_collections_populated(history, mapping_params.param_template)
  File "/opt/galaxy/lib/galaxy/tools/execute.py", line 260, in ensure_implicit_collections_populated
    self.precreate_output_collections(history, params)
  File "/opt/galaxy/lib/galaxy/tools/execute.py", line 278, in precreate_output_collections
    effective_structure = self._mapped_output_structure(trans, output)
  File "/opt/galaxy/lib/galaxy/tools/execute.py", line 244, in _mapped_output_structure
    output_structure = tool_output_to_structure(self.sliced_input_collection_structure, tool_output, collections_manager)
  File "/opt/galaxy/lib/galaxy/dataset_collections/structure.py", line 183, in tool_output_to_structure
    tree = get_sliced_input_collection_structure(structured_like)
  File "/opt/galaxy/lib/galaxy/tools/execute.py", line 214, in sliced_input_collection_structure
    input_collection = self.example_params[input_name]
KeyError: 'input_paired'
127.0.0.1 - - [16/May/2018:23:42:05 +0100] "POST /api/tools HTTP/1.1" 500 - "http://127.0.0.1:8080/?tool_id=toolshed.g2.bx.psu.edu%2Frepos%2Fiuc%2Fsickle%2Fsickle%2F1.33.1&version=1.33.1&__identifer=eqdjaao7u85" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0"
```

The relevant parts from the tool XML:
```xml
    <inputs>
        <conditional name="readtype">
            <param name="single_or_paired" type="select" label="Single-end or paired-end reads?" help="Note: Sickle will infer the quality type of the file from its datatype.  I.e., if the datatype is fastqsanger, then the quality type is sanger. The default is fastqsanger.">
                <option value="se" selected="true">Single-end</option>
                <option value="pe_combo">Paired-end (one interleaved input file)</option>
                <option value="pe_sep">Paired-end (two separate input files)</option>
                <option value="pe_collection">Paired-end (as collection)</option>
            </param>
...
            <when value="pe_collection">
                <param format="fastq,fastq.gz" name="input_paired" type="data_collection" collection_type="paired" label="Paired-end FASTQ reads as paired collection" />
            </when>
        </conditional>
...
    <outputs>
...
        <collection name="output_paired_coll" type="paired" structured_like="input_paired" inherit_format="true" label="Paired-end output of ${tool.name} on ${on_string}">
            <filter>readtype['single_or_paired'] == 'pe_collection'</filter>
        </collection>
...
    </outputs>
```

A test for this was introduced in https://github.com/galaxyproject/galaxy/pull/3209 , but in `test/functional/tools/paired_collection_map_over_structured_like.xml` the paired collection input is not inside of a conditional.

Ping @jmchilton

**Comment by jmchilton** [comment]
I'll write up a new test case for this. Thanks for the detailed report.

**Comment by jmchilton** [comment]
I've reproduced this with a test case and I can imagine a fix - but it is a bit of a hack I think. In https://github.com/galaxyproject/galaxy/issues/5756 @mvdbeek realized that ``default_identifier_source`` requires a fully qualified parameter - same case here now it seems - this would work I think if the tool had ``structured_like="readtype|input_paired"``. Probably just updating the tool and moving on isn't good because reproducibility. I guess my question is should I limit the hack to older tool profile versions to encourage fully qualified inputs - or do we like the magic that finds the unqualified input magically in the input tree. I sort of hate the magic because it is ambiguous - input_paired may appear multiple places in the inputs tree if the parameter is in multiple conditionals or repeats? I feel like maybe people don't agree too much with me though since people still write tool tests using a very similar syntax to this instead of writing fully qualified parameter descriptions.

**Comment by mvdbeek** [comment]
Yeah, I intuitively went with "let's use the fully qualified option", but admittedly I reviewed the original PR and didn't even think about it, I only figured it out by checking what's going on in the debugger. I think limiting this behavior to a certain profile version is a good idea.

**Comment by nsoranzo** [comment]
I already tried `structured_like="readtype|input_paired"` and it didn't work.

**Comment by jmchilton** [comment]
> I already tried structured_like="readtype|input_paired" and it didn't work.

Indeed - just getting there - example_params is not flat:

```
{'cond': {'input1': <galaxy.model.DatasetCollectionElement object at 0x112288b90>, 'cond_param': u'paired', '__current_case__': 0}}
```

Galaxy tool state makes me constantly sad.

**Comment by nsoranzo** [comment]
I'm OK with requiring the fully qualified input in `structured_like` if that simplifies the implementation, I've searched in tools-iuc, tools-devteam, tools-galaxyp, earlham-galaxytools, galaxy_blast and bgruening/galaxytools repos and `sickle` is the only tool using this feature.

**Comment by jmchilton** [comment]
So I have the fix and requiring it only eliminates three lines of code - to me it is less about simplifying the implementation than building a tool spec that is cohesive and makes sense. I'll keep the three lines and wrap them in a profile < "18.09" wrapper gate I think. Thanks for searching for that though, it is good to understand the scope of the problem.

**Comment by nsoranzo** [comment]
Fixed by #6162.
